napari/_qt/dialogs/qt_reader_dialog.py

```diff
@@ -88,61 +94,146 @@ def _get_persist_choice(self):
+    ----------
+    _paths : list[str]
+        list of paths to open, as strings
+    viewer : QtViewer
+        QtViewer to associate dialog with
+    stack : bool
+        True if list of paths should be stacked, otherwise False
+    plugin_name : str | None
+        name of plugin already tried, if any
+    error : ReaderPluginError | None
+        previous error raised in the process of opening
+    """
+    _path = paths[0]
+    readers = prepare_remaining_readers(_path, plugin_name, error)
+    # TODO: fix tests
+    error_message = str(error) if error else ''
+
+    _, extension = os.path.splitext(_path)
+    readerDialog = QtReaderDialog(
+        parent=qt_viewer,
+        pth=_path,
+        extension=extension,
+        error_message=error_message,
+        readers=readers,
+    )
+    display_name, persist = readerDialog.get_user_choices()
+    if display_name:
+        open_with_dialog_choices(
+            display_name, persist, extension, readers, paths, stack, qt_viewer
+        )
 
-    If there is just one reader and no error message, dialog
-    is not shown. Otherwise, launch dialog and ask user for
-    plugin choice and whether setting is persisted.
 
-    Returns None if user cancels on dialog.
+def prepare_remaining_readers(
+    _path: str,
+    plugin_name: Optional[str] = None,
+    error: Optional[ReaderPluginError] = None,
+):
+    """Remove tried plugin from readers and raise error if no readers remain.
 
     Parameters
     ----------
-    readerDialog : QtReaderDialog or MockQtReaderDialog
-        reader dialog to use for choices from the user
-    readers : str
-        Dictionary of display-name:plugin-name of potential readers for file
-    has_errored : bool
-        True when we've tried to read this file and failed, otherwise False
+    _path : str
+        path to open
+    plugin_name : str | None
+        name of plugin previously tried, if any
+    error : ReaderPluginError | None
+        previous error raised in the process of opening
 
     Returns
     -------
-    display_name: str
-        Display name of the chosen plugin
-    persist_choice: bool
-        Whether to persist the chosen plugin choice or not
+    readers: Dict[str, str]
+        remaining readers to present to user
 
+    Raises
+    ------
+    error
+        raises previous error if no readers are left to try
     """
-    display_name = ''
-    persist_choice = False
-
-    # if we have just one reader and no errors from existing settings
-    if len(readers) == 1 and not has_errored:
-        # no need to open the dialog, just get the reader choice
-        display_name = next(iter(readers.keys()))
-        return display_name, persist_choice
+    readers = get_potential_readers(_path)
+    # remove plugin we already tried e.g. prefered plugin
+    if plugin_name in readers:
+        del readers[plugin_name]
+    # if there's no other readers left, raise the exception
+    if not readers and error:
+        raise error
```

Comment:
I still don't like this pattern, but in this case I don't see a clear way to fit it into a try/except. Having said this, I think this *has* to be a ReaderPluginError (?), in which case that should be made clear in the docstring, *and* potentially, this should be a raise NewError(message) from error? I don't know what kind of stack trace we are going to get otherwise, but I suspect it is a less helpful one. Example (if I've followed the logic correctly): "ReaderPluginError: tried to read path {path} with plugin {plugin}, because it was associated with that path/because it is the only plugin capable of reading that path, but it gave an error. Try associating a different plugin or installing a different plugin for this kind of file."